Refact: typos and shell improvements

taimoorgit opened the pull request:
PR is based on: https://github.com/moogacs/nilaway-plugin/pull/1 and so it includes changes from that PR. 

* [Fix shellcheck warnings and errors](https://www.shellcheck.net/) from build.sh, build-local.sh
  * These edits are just quoting variables
  * Fixed a few lines that were too long to read
  * Also add `set -euo pipefail` so the build can end faster if something went wrong
  * Add `usage` that gets printed out reminding user they need to pass the version as an arg to ./build
* Fix some typos and ~~capitalization~~
* Fix line that outputs go version - that go version being printed out is the one that was used to build golangci-lint, not the one being used to build the plugin, which is confusing. 
  * It's good to know if I am building using Go 1.21 but my golangci-lint was actually built with Go 1.22. it might not matter but is useful for the user to know about the mismatch.
  * Added the color yellow to make this more visible.

Commit:
use -trimpath with go build

Commit:
fix shellcheck warnings, caps and typos

taimoorgit commented on build.sh:
```diff
@@ -1,13 +1,13 @@
 #!/bin/bash
 
-echo $1
```
Warning for this line and others was thrown [by `shellcheck`](https://www.shellcheck.net/) about globbing/splitting.

Commit:
highlight diff btwn go versions on build

taimoorgit commented on build.sh:
```diff
@@ -1,13 +1,22 @@
+go_version=$(grep -oE '^go [0-9]+\.[0-9].+' <<< "$output" | cut -d ' ' -f2)
+
+local_go_version=$(go version | grep -oE '[0-9]+\.[0-9]+\.[0-9]+')
+if [[ "$go_version" != "$local_go_version" ]]; then
```
Reason for adding this if: it would be good to know that there is a version mismatch, i.e. I am using Go 1.21 and golangci-lint v1.56.X is using Go 1.22. 

It might not make a difference so this just outputs error messages and does not stop the build.

taimoorgit commented on local_build.sh:
```diff
@@ -1,16 +1,25 @@
+go mod edit -replace golang.org/x/tools=golang.org/x/tools@v"$tools_version"
 go mod tidy
 
-echo building nilaway.so with GOARCH=$GOARCH GOOS=$GOOS tools_version=$tools_version go_version=$go_version ...
```
Original line is a little confusing since the GOARCH, GOOS variables are the same as on the user's PC but the `go_version` printed out here has nothing to do with the go version being used to build the plugin.

Commit:
set -euo pipefail and color

moogacs commented on local_build.sh:
```diff
@@ -1,16 +1,28 @@
+  echo -e "${YELLOW}warning: local go version ($local_go_version) != go version used to build $(which golangci-lint) ($go_version)${NC}" &&
+  echo -e "${YELLOW}continuing build like normal ...${NC}"
+
+echo "$tools_version"
```
**suggestion (non-blocking):** what do you think if we add here `rm go.*` before go mod init so incase of the go mod files exists it would be replaced

moogacs commented on build.sh:
```diff
@@ -1,13 +1,26 @@
+  echo -e "${YELLOW}warning: local go version ($local_go_version) != go version in go.mod from golangci-lint@$1  ($go_version)${NC}" &&
+  echo -e "${YELLOW}continuing build like normal ...${NC}"
 
 go mod init github.com/nilaway-plugin
```
**suggestion (non-blocking):**  what do you think if we add here rm go.* before go mod init so incase of the go mod files exists it would be replaced

moogacs commented on build.sh:
**note (if-minor):** i think we don't need -e after echo across the whole file

moogacs commented on local_build.sh:
**note (if-minor):** i think we don't need -e after echo across the whole file

moogacs reviewed (commented):
Thanks for taking the time and improve the script ;) looks nice. Just left some suggestions and notes, also if you can update the PR title to something more descriptive. something like :
`refact: typos and shell improvements`

Commit:
add requested changes

taimoorgit commented:
> Thanks for taking the time and improve the script ;) looks nice. Just left some suggestions and notes, also if you can update the PR title to something more descriptive. something like : `refact: typos and shell improvements`

Sure! 

I just updated the PR title and added a commit,

* Check if `$GITHUB_WORKSPACE` is set (it should be if running in actions afaict. But just in case someone tries to use the script locally, e.g. in your own environment to test it out like I am doing, it's useful to have this check...)
* Use an actual `if` for the version check - easier to read.
* Use `echo -e` only for the version check output, everything else can be normal echo.
* Replaced the `echo building nilaway.so with <long text>` with just using a `set -x` call so that the command is just shown in the output, we don't have to have that long line. Also makes it more clear as to what specific command is being run.
* Added your suggestion to do `rm go.*` - added code at the top of both files to do "cd into dirname first so we don't delete anything unexpected".
* Removed some unnecessary `&&` now that `set -e` is enabled.

Let me know what you think. Also -- it might just be easier for us to merge these 2 build scripts into one, since so much of the code is repeated. I can look into doing that.

taimoorgit commented:
```
$ ./local_build.sh
+ go mod init github.com/nilaway-plugin
go: creating new go.mod: module github.com/nilaway-plugin
go: to add module requirements and sums:
        go mod tidy
+ go mod tidy
go: finding module for package golang.org/x/tools/go/analysis
go: finding module for package go.uber.org/nilaway
go: found go.uber.org/nilaway in go.uber.org/nilaway v0.0.0-20240224031343-67945fb5199f
go: found golang.org/x/tools/go/analysis in golang.org/x/tools v0.18.0
+ go mod edit -replace golang.org/x/tools=golang.org/x/tools@v0.18.0
+ go mod tidy
+ GOARCH=arm64
+ GOOS=darwin
+ go build -buildmode=plugin -trimpath -o nilaway.so ./plugin/nilaway.go
+ set +x
nilaway.so is built compatible with local golangci-lint, please move it under your project directory and make sure to copy the custom settings to your .golangci.yml file.

$ ./build.sh
$GITHUB_WORKSPACE is not set.

$ GITHUB_WORKSPACE="." ./build.sh latest
warning: local go version (1.22.0) != go version in go.mod from golangci-lint@latest  (1.21)
continuing build like normal ...
go: creating new go.mod: module github.com/nilaway-plugin
go: to add module requirements and sums:
        go mod tidy
go: finding module for package golang.org/x/tools/go/analysis
go: finding module for package go.uber.org/nilaway
go: found go.uber.org/nilaway in go.uber.org/nilaway v0.0.0-20240224031343-67945fb5199f
go: found golang.org/x/tools/go/analysis in golang.org/x/tools v0.18.0
nilaway.so is built compatible with golangci-lint@latest.
```

moogacs commented:
👍🏼 I like the idea of combining both in 1, lets then close this PR and work on the other one  https://github.com/moogacs/nilaway-plugin/pull/4
